### PR TITLE
Brian/better tx evaluation error

### LINKED ISSUE #732: Tx evaluation error parsing and transport
Currently, properly constructed Tx may fail to balance with a very uninformative error that is being thrown during Ogmios EvaluateTx response parsing.
It'd be better to either improve the parsing algorithm so that it can parse all ogmios responses or provide a way to inspect unparsed faulty responses to CTL clients. Such responses may contain details on how validator failed to evaluate the transaction, which is important from a debugging perspective of things and instead users get "balance errors" which is just confusing.

### PR DESCRIPTION
**Pull request by Geometer1729**
Closes #732 .

### Pre-review checklist

- [x] All code has been formatted using our config (`make format` for Purescript, `nixpkgs-fmt` for Nix)
- [x] All Purescript imports are explicit, including constructors
- [ ] The integration and unit tests have been run (`npm run test`) locally
- [x] The changelog has been updated under the `## Unreleased` header, using the appropriate sub-headings (`### Added`, `### Removed`, `### Fixed`), and the links to the appropriate issues/PRs have been included

### CODE REVIEW

**Comment by Geometer1729** [comment]
`npm run test` is giving me
```
16/17 tests passed

[error] Running failed; exit code: 1
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! cardano-transaction-lib@2.0.0-alpha integration-test: `spago run --main Test.Integration`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the cardano-transaction-lib@2.0.0-alpha integration-test script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
npm WARN Local package.json exists, but node_modules missing, did you mean to install?
```
but `spago test` works. I'm not sure what to make of that.

**Comment by Geometer1729** [comment]
I'm checking off the changelog thing because I think it's fine that I haven't put anything in the changelog for this.

**Comment by jy14898** [comment]
Apologies for the duplicated work, I didn't yet create a PR for the work I've been doing on this issue, so I've made it available here: https://github.com/Plutonomicon/cardano-transaction-lib/pull/832

File: src/BalanceTx/BalanceTx.purs [hunk 1]
```diff
@@ -380,27 +390,26 @@ reattachDatumsAndRedeemers
...
       let
-        xsWrapped = Ogmios.TxEvaluationR (Map.fromFoldable xs)
         ixMaybe = flip Array.findIndex rs $ \(Redeemer rdmr /\ _) ->
           rdmr.tag == ptr.redeemerTag
             && rdmr.index == Natural.toBigInt ptr.redeemerIndex
       in
-        ixMaybe # maybe (setRdmrsExecutionUnits rs xsWrapped) \ix ->
-          flip setRdmrsExecutionUnits xsWrapped $
+        ixMaybe # maybe (setRdmrsExecutionUnits rs (Map.fromFoldable xs)) \ix ->
+          flip setRdmrsExecutionUnits (Map.fromFoldable xs) $
```
Comment (klntsky) [inline src/BalanceTx/BalanceTx.purs:419]: Do not call fromFoldable twice, introduce a variable.

File: src/BalanceTx/BalanceTx.purs [hunk 2]
```diff
@@ -380,27 +390,26 @@ reattachDatumsAndRedeemers
       # _witnessSet <<< _redeemers ?~ map fst redeemersTxIns
 
 updateTxExecutionUnits
-  :: UnattachedUnbalancedTx -> Ogmios.TxEvaluationR -> UnattachedUnbalancedTx
+  :: UnattachedUnbalancedTx -> Ogmios.CostMap -> UnattachedUnbalancedTx
 updateTxExecutionUnits unattachedTx rdmrPtrExUnitsList =
```
Comment (klntsky) [inline src/BalanceTx/BalanceTx.purs:401]: please change the argument name also.

File: src/QueryM/Ogmios.purs [hunk 3]
```diff
@@ -466,20 +467,27 @@ type RedeemerPointer = { redeemerTag :: RedeemerTag, redeemerIndex :: Natural }
...
 instance DecodeAeson TxEvaluationR where
-  decodeAeson = aesonObject $ \obj -> do
-    rdmrPtrExUnitsList :: Array (String /\ Aeson) <-
-      ForeignObject.toUnfoldable <$> getField obj "EvaluationResult"
-    TxEvaluationR <<< Map.fromFoldable <$>
-      traverse decodeRdmrPtrExUnitsItem rdmrPtrExUnitsList
+  decodeAeson = aesonObject $ \obj ->
+    (TxEvaluationFailure <$> getField obj "EvaluationFailure") <|>
+      -- Putting the falure case first gives better errors in some cases
+      ( do
```
Comment (klntsky) [inline src/QueryM/Ogmios.purs:485]: Parentheses are not required

**Review by klntsky — CHANGES_REQUESTED** [review]

**Pull request closed without merging**